bookwyrm/tests/views/preferences/test_export.py

```diff
@@ -66,7 +69,7 @@ def test_export_file(self, *_):
         # pylint: disable=line-too-long
         self.assertEqual(
             export.content,
-            b"title,author_text,remote_id,openlibrary_key,inventaire_id,librarything_key,goodreads_key,bnf_id,viaf,wikidata,asin,aasin,isfdb,isbn_10,isbn_13,oclc_number,start_date,finish_date,stopped_date,rating,review_name,review_cw,review_content\r\nTest Book,,"
+            b"title,author_text,remote_id,openlibrary_key,inventaire_id,librarything_key,goodreads_key,bnf_id,viaf,wikidata,asin,aasin,isfdb,isbn_10,isbn_13,oclc_number,start_date,finish_date,stopped_date,rating,review_name,review_cw,review_content,review_published,shelf,shelf_name,shelf_date\r\nTest Book,,"
             + self.book.remote_id.encode("utf-8")
-            + b",,,,,beep,,,,,,123456789X,9781234567890,,,,,,,,\r\n",
+            + b",,,,,beep,,,,,,123456789X,9781234567890,,,,,,,,,,,,\r\n",
```

Comment:
The test is failing because your code adds the `shelf`, `shelf_name` and `shelf_date` to each row but you're asserting these will be empty values.

The first two will be `to-read` and `To Read`, but the `shelf_date` is trickier in this test because you need to assign a date dynamically, since it will change whenever the test is run. As a bonus level of complexity we will have to do this by adding a variable to a byte string, so we can't use an f-string statement because they don't work for byte-strings. So it's not very surprising that you are having trouble with this!

So first of all we need to assign a variable name to the ShelfBook so we can reference it to get the date:

```py
shelfbook = models.ShelfBook.objects.create(
    shelf=self.local_user.shelf_set.first(),
    user=self.local_user,
    book=self.book,
)
```
Then just above the assert statement we need to add a variable where we turn the date into a byte string:

```py
book_date = str.encode(f"{shelfbook.shelved_date.date()}")
```

And then in the final line of the expected output we need to add the shelf names and a placeholder for the date (`%b`), which we assign with `%`:
```py
+ b",,,,,beep,,,,,,123456789X,9781234567890,,,,,,,,,,to-read,To Read,%b\r\n" % book_date,
```

Reply:
Thank you very much for this explanation. I understand where the problem comes, but I've realised I do not understand where the test data comes from! 🤦 (i.e. how do you know that the shelf is `to-read` and not any other or blank?)

I can see line 37 that seems to create some data for some fields, but A) data from other fields (such as shelve) is not there; and B) I can see that the contents from `bnf_id` and `isbn_13` are in the csv, but not the others (i.e. `title` or `remote_id`). 

Probably this shows how lost I am with this!

Reply:
Don't worry, when I first started working on BookWyrm I found the tests completely bamboozling!

The test data is created, as you identified, within the `setUpTestData`. This creates a bunch of data and attaches it to `self`, which refers to the class: in this case `ExportViews`. This allows us to re-use that test data in each test within the test case, and the data is reset for each test. It's really handy if you want to test a range of different scenarios with basically the same test data.

### Where is the title?

The way this test is written is slightly confusing because the beginning of the second row (i.e. the first row after the csv headers) is actually at the _end_ of line 72. This threw me for a moment as well. So you can see that the title is indeed the first piece of data in the first actual row:
```py
...shelf_date\r\nTest Book,,"
```
I think this is so that the remote_id can be added dynamically, though as you can see this is done differently to my suggestion above. You could make it a bit easier for everyone to understand by reformatting like this:

```py
self.assertEqual(
    export.content,
    b"title,author_text,remote_id,openlibrary_key,inventaire_id,librarything_key,goodreads_key,bnf_id,viaf,wikidata,asin,aasin,isfdb,isbn_10,isbn_13,oclc_number,start_date,finish_date,stopped_date,rating,review_name,review_cw,review_content,review_published,shelf,shelf_name,shelf_date\r\n"
    + b"Test Book,,%b,,,,,beep,,,,,,123456789X,9781234567890,,,,,,,,,,to-read,To Read,%b\r\n" % (self.book.remote_id.encode("utf-8"), book_date)
)
```
### How do we know what the shelf is called?

Ok so to be perfectly honest I knew this because I looked at the failed test output! But there's a good explanation:

A `ShelfBook` _must_ have a shelf, a user, and a book. If you try to create one without one of those things you'll get an error. In your test we're creating a shelfbook and pass in the shelf like this:

```py
shelf=self.local_user.shelf_set.first(),
```

Which means "get all the `Shelf` instances belonging to `self.local_user`, and select the first one".

When a `User` is created, the `save` function triggers [a `create_shelves` function](https://github.com/bookwyrm-social/bookwyrm/blob/b04ebe397b78a7fb4a6d3c69f5f9cc134d99e2e6/bookwyrm/models/user.py#L465) and the first shelf in the list is:

```py
{
  "name": "To Read",
  "identifier": "to-read",
},
```

Since `create_shelves` will run as soon as a user is created, that means we know that the first shelf in any user's list of shelves will always be "To Read" unless it's sorted a different way.

### Where does the remote id come from?

The `remote_id` for `self_book` _looks like_ it's created in `models.Edition.objects.create`, however this is a gotcha, because the remote id that is passed in gets changed to `origin_id` when the Edition is saved:

https://github.com/bookwyrm-social/bookwyrm/blob/b04ebe397b78a7fb4a6d3c69f5f9cc134d99e2e6/bookwyrm/models/book.py#L95C1-L102C45

...so that's why we have to pass it in as a variable to the expected CSV output.

Apologies, that's all pretty in the weeds so if it doesn't make sense, let me know.